Code/Geometry/Trajectory.cpp

```diff
@@ -0,0 +1,210 @@
+//
+// Copyright (C) 2003-2016 Paolo Tosco
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+
+#include <RDGeneral/Invariant.h>
+#include <RDGeneral/BadFileException.h>
+#include <fstream>
+#include <sstream>
+#include <set>
+#include "Trajectory.h"
+
+namespace RDGeom {
+
+Snapshot::Snapshot(double *pos, double energy, bool owner) :
+  d_trajectory(NULL),
+  d_owner(owner),
+  d_energy(energy),
+  d_pos(pos)
+{
+}
+
+Point2D Snapshot::getPoint2D(unsigned int pointNum) const {
+  PRECONDITION(d_pos, "pos must not be NULL");
+  PRECONDITION(trajectory()->dimension() == 2, "d_dimension must be == 2");
+  unsigned int i = pointNum * trajectory()->dimension();
+  return Point2D(d_pos[i], d_pos[i + 1]);
+}
+
+Point3D Snapshot::getPoint3D(unsigned int pointNum) const {
+  PRECONDITION(trajectory()->dimension() >= 2, "d_dimension must be >= 2");
+  unsigned int i = pointNum * trajectory()->dimension();
+  return (d_pos ? Point3D(d_pos[i], d_pos[i + 1],
+          ((trajectory()->dimension() == 3) ? d_pos[i + 2] : 0.0)) : Point3D());
+}
+
+void Snapshot::setTrajectory(const Trajectory *traj) {
+  PRECONDITION(traj, "traj must not be NULL");
+  d_trajectory = traj;
+}
+
+void Snapshot::freePos() {
+  delete d_pos;
+  d_pos = NULL;
+}
+
+Trajectory::Trajectory(unsigned int dimension, unsigned int numPoints) :
+  d_dimension(dimension),
+  d_numPoints(numPoints)
+{
+}
+
+Trajectory::~Trajectory() {
+  for (std::vector<Snapshot>::iterator it = d_snapshotVect.begin();
+    it != d_snapshotVect.end(); ++it) {
+    if (it->getOwner())
+      it->freePos();
+  }
+}
+
+unsigned int Trajectory::addSnapshot(Snapshot s) {
```

Comment:
This looks dangerous, two snapshots could be "owners" of the same memory.

Perhaps use a shared_ptr to wrap the memory location as opposed to having an "owner?"